opt: infer filters from equivalent columns

rytaft opened the pull request:
This PR includes 3 commits:

1. Update logical properties with additional functional dependencies to make it possible to infer equivalent columns in Join rules.
2. Update tests based on the updated functional dependencies.
3. Add rules for filter inference and push-down, in which new filters are inferred based on equivalent columns.

rytaft committed:
opt: Add functional dependencies to Scalar props

This commit adds functional dependencies (FDs) to Scalar props so that
FDs can be stored in the logical properties for Filters expressions.
FDs that can be inferred from Filters expressions include:
 - Constant column FDs such as ()-->(1,2) from conjuncts such as
   x = 5 AND y = 10.
 - Equivalent column FDs such as (1)==(2), (2)==(1) from conjuncts such
   as x = y.

It is useful to calculate FDs on Filters expressions, because it allows
additional filters to be inferred for push-down. For example, consider
the query:

  SELECT * FROM a, b WHERE a.x = b.x AND a.x > 5;

By adding the equivalency FD for a.x = b.x, we can infer an additional
filter, b.x > 5. This allows us to rewrite the query as:

  SELECT * FROM (SELECT * FROM a WHERE a.x > 5) AS a,
    (SELECT * FROM b WHERE b.x > 5) AS b WHERE a.x = b.x;

In addition, this commit also updates the logicalPropsBuilder to
use the Filters FDs and Constraints when calculating the FDs for
constrained Scan expressions, joins, and Select expressions. It also
improves identification of the not null columns using Constraints.

Release note: None

rytaft committed:
opt: update tests for new FD field in scalar props

Update the test cases to display functional dependencies
for the Filters expression, as well as updated FDs for scans,
selects, and joins.

Release note: None

rytaft committed:
opt: infer filters from equivalent columns

This commit adds several rules to infer filters from
equivalent columns for the purpose of filter push-down.

The rules are:

1. PushSelectIntoJoinLeftAndRight

   This rule pushes a filter into both the left and right sides of a join
   if it can be mapped to use the columns of both sides. For example,
   consider this query:

   SELECT * FROM a, b WHERE a.x=b.x AND a.y=b.y AND a.x + b.y < 5

   In this case, we can map a.x + b.y < 5 to both sides based on the
   equality filters a.x=b.x AND a.y=b.y. For the left side, we can map it to
   a.x + a.y < 5, and for the right side, we can map it to b.x + b.y < 5.
   Given this mapping, we can safely push the filter down to both sides and
   remove it from the Select filters list.

   Note that we can apply this rule to any join except for semi- and
   anti-joins because we are filtering *both* sides of the join. If we can
   only push down to one side, this rule does not apply. For example, if we
   push a filter to only the right side of a left join, that will not reduce
   the number of output rows but will result in null columns in the output.

2. MapSelectIntoJoinLeft

   This rule maps a filter that is not bound by the left side of
   the join to use the columns from the left side. This will allow
   the filter to be pushed down by the PushSelectIntoJoinLeft rule.
   For example, consider this query:

   SELECT * FROM a, b WHERE a.x = b.x AND b.x + a.y < 5

   In this case, we can map b.x + a.y < 5 to the left side by replacing b.x
   with the equivalent column a.x.

   This rule only applies to InnerJoins and LeftJoins, in cases where it
   is not possible to map the filter to both sides. If the filter can be
   mapped to both sides, it will be handled by PushSelectIntoJoinLeftAndRight.

3. MapSelectIntoJoinRight

   This rule is symmetric with MapSelectIntoJoinLeft. It maps Select filter
   conditions to use columns from the right side of the join rather than the
   left side. It applies to InnerJoins and RightJoins.

4. MergeSelectEqualityJoin

   This rule is a specialization of an existing rule, MergeSelectInnerJoin,
   which merges a Select operator with an InnerJoin input by AND'ing the
   filter conditions of each and creating a new InnerJoin with that On
   condition. MergeSelectEqualityJoin is specialized for the case where the
   InnerJoin has an equality filter, so combining the filters could allow for
   new filter inference based on equivalent columns.

5. PushFilterIntoJoinLeftAndRight

   This rule is similar to PushSelectIntoJoinLeftAndRight, but it applies to
   filter conditions in the ON clause rather than the WHERE clause. In this
   case, it is only safe to apply the rule to InnerJoins and SemiJoins.

6. MapFilterIntoJoinLeft

   This rule is similar to MapSelectIntoJoinLeft, but it applies to filter
   conditions in the ON clause rather than the WHERE clause. It applies to
   InnerJoins, RightJoins, and SemiJoins.

7. MapFilterIntoJoinRight

   This rule is symmetric with MapFilterIntoJoinLeft. It maps Join filter
   conditions to use columns from the right side of the join rather than the
   left side. It applies to InnerJoins, LeftJoins, SemiJoins, and AntiJoins.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26937)
<!-- Reviewable:end -->

rytaft commented:
Based on offline discussion with Radu, I modified the code in `opt/memo/logical_props_builder.go` to only add equivalences for columns which do not have a composite key encoding.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LFsgybnDmjr4F7mdkEU:b-60vpwu)*
<!-- Sent from Reviewable.io -->

andy-kimball commented on pkg/sql/opt/norm/rules/join.opt:
```diff
@@ -37,6 +37,83 @@
+(InnerJoin | InnerJoinApply | SemiJoin | SemiJoinApply
+    $left:*
+    $right:*
+    $filters:(Filters $list:[ ... $condition:* & ^(Eq (Variable) (Variable)) & (CanMap $filters $condition $left) & (CanMap $filters $condition $right) ... ])
```
Add line breaks to make this more readable. I'd typically do this:
```
(InnerJoin | InnerJoinApply | SemiJoin | SemiJoinApply
    $left:*
    $right:*
    $filters:(Filters
        $list:[
            ...
            $condition:* &
                ^(Eq (Variable) (Variable)) &
                (CanMap $filters $condition $left) &
                (CanMap $filters $condition $right)
            ...
        ])
)
```

andy-kimball commented on pkg/sql/opt/norm/rules/join.opt:
```diff
@@ -37,6 +37,83 @@
+((OpName)
+    (Select
+        $left
+        (Filters (InternSingletonList (Map $filters $condition $left)))
```
I'd use the list constructor syntax here:
```
(Filter [(Map $filters $condition $left)])
```

andy-kimball commented on pkg/sql/opt/norm/rules/join.opt:
```diff
@@ -37,6 +37,83 @@
+(InnerJoin | InnerJoinApply | RightJoin | RightJoinApply | SemiJoin | SemiJoinApply
+    $left:*
+    $right:*
+    $filters:(Filters $list:[ ... $condition:* & ^(Eq (Variable) (Variable)) & ^(IsBoundBy $condition $left) & (CanMap $filters $condition $left) ... ])
```
Same line break comment as above.

andy-kimball commented on pkg/sql/opt/norm/rules/join.opt:
```diff
@@ -37,6 +37,83 @@
+ SemiJoin | SemiJoinApply | AntiJoin | AntiJoinApply
+    $left:*
+    $right:*
+    $filters:(Filters $list:[ ... $condition:* & ^(Eq (Variable) (Variable)) & ^(IsBoundBy $condition $right) & (CanMap $filters $condition $right) ... ])
```
Same line break comment as above.

andy-kimball commented on pkg/sql/opt/norm/rules/select.opt:
```diff
@@ -82,6 +82,131 @@
+        $right:*
+        $on:*
+    )
+    $filters:(Filters $list:[ ... $condition:^(Eq (Variable) (Variable)) & (CanMap $filters $condition $left) & (CanMap $filters $condition $right) ... ])
```
Same comments as above (line break + use list constructor syntax).

andy-kimball commented on pkg/sql/opt/norm/testdata/rules/join:
```diff
@@ -321,6 +321,390 @@ anti-join
+      └── filters [type=bool, outer=(1,2,6,11), constraints=(/1: (/NULL - ]; /6: (/NULL - ]), fd=(1)==(6), (6)==(1)]
+           ├── a.k = b.x [type=bool, outer=(1,6), constraints=(/1: (/NULL - ]; /6: (/NULL - ])]
+           └── (b.x * a.i) = min [type=bool, outer=(2,6,11)]
+
```
How about a test case that involves equality conditions that only involve columns from one side or the other? The idea is to try and confuse the rules with a non-join equality.

andy-kimball reviewed (approved):
LGTM, with some smallish comments. Nice work!

RaduBerinde commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/norm/custom_funcs.go, line 435 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFxEpio3lXOMMo9UCdd:-LFxEpip8l8yFo-u7eBZ:b-bn7q0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L435)):*
> ```Go
> 		eqCols := fd.ComputeEquivClosure(util.MakeFastIntSet(srcCol))
> 		eqDstCols := eqCols.Intersection(c.OutputCols(dst))
> 		if eqDstCols.Contains(srcCol) {
> ```

Should we check that the types of the columns are the same? (either here or when we build the FDs) E.g. we may have `a = b` where `a` is float and `b` is int, and it's not always ok to replace one with the other. Some expressions might even become invalid (e.g. `cos(a)` is valid but `cos(b)` is not).

---

*[pkg/sql/opt/norm/rules/join.opt, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFscxbk4VMJZMpXL1vz:-LFscxbk4VMJZMpXL1w-:b-mgym2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/978fe10f1f8b875cfbf135f21eaf18cffc56e87f/pkg/sql/opt/norm/rules/join.opt#L57)):*
> ```
>     $left:*
>     $right:*
>     $filters:(Filters $list:[ ... $condition:* & ^(Eq (Variable) (Variable)) & (CanMap $filters $condition $left) & (CanMap $filters $condition $right) ... ])
> ```

On the `(Eq Var Var)` exception: we could have something like `left.a=right.a AND left.a=right.b` and it would be nice to push that down as `right.a=right.b`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937)*
<!-- Sent from Reviewable.io -->

rytaft committed:
opt: add rules to simplify joins wrapped in null-rejecting filter

This commit adds two rules to simplify joins when the are wrapped in
a Select with a null-rejecting filter. The rules are:

1. SimplifySelectLeftJoin reduces a LeftJoin operator to an InnerJoin operator
   (or a FullJoin to a RightJoin) when there is a null-rejecting filter on any
   column from the right side. The effect of the null-rejecting filter is that
   output rows with all NULL values on the right side created by the left (or
   full) join are eliminated, making the join equivalent to an inner (or right)
   join. For example:

     SELECT * FROM a LEFT OUTER JOIN b ON a.x = b.x WHERE b.y < 5

   can be reduced to:

     SELECT * FROM a INNER JOIN b ON a.x = b.x WHERE b.y < 5

   since b.y < 5 is a null-rejecting filter on the right side.

2. SimplifySelectRightJoin is symmetric with SimplifySelectLeftJoin. It reduces
   a RightJoin operator to an InnerJoin operator (or a FullJoin to a LeftJoin)
   when there is a null-rejecting filter on any column from the left side.

These rules also allow several of the filter push-down rules created by a
previous commit in select.opt to be eliminated, since those rules were designed
to push down filters for queries such as:

  SELECT * FROM a OUTER JOIN b ON true WHERE a.x=b.x AND a.x < 5

With the new rules, this query will always be converted to an inner join, and
the Select filter can be merged with the ON filter. Filter push-down rules in
join.opt can then perform push-down.

Release note: None

rytaft commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/memo/expr_view.go, line 442 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwW4zOBlFn2a61R-r5-442:-LG1VRHW21I_qKrPe9S5:b-376vyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/25e8d8bdc7f39651e32cf1c7bf8bd46a582a983b/pkg/sql/opt/memo/expr_view.go#L442)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Need `|| hasFuncDeps` here
</blockquote></details>

Done. Good catch!

---

*[pkg/sql/opt/norm/custom_funcs.go, line 394 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis2-r7-394:-LG1VSJu4jgc0xPfctFc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L394)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It'd be slightly better to use the `Next` pattern so that you could immediately return `false` as soon as you find a column that can't be mapped.
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/custom_funcs.go, line 421 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis2-r7-421:-LG1VTCUDbmHpoGmglpp:bpm4gyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L421)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

If it's already known to be possible, then this is a dead path rather than a fast path. : )
</blockquote></details>

This path is actually used by the `PushFilterIntoJoinLeftAndRight` rule -- it's possible that a condition is already bound by one side, but the rule will create a new filter bound by the other side and push both down together.

---

*[pkg/sql/opt/norm/custom_funcs.go, line 435 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFxEpio3lXOMMo9UCdd:-LG1S_K-8OYNOXLsrLfe:bcq2fwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L435)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Should we check that the types of the columns are the same? (either here or when we build the FDs) E.g. we may have `a = b` where `a` is float and `b` is int, and it's not always ok to replace one with the other. Some expressions might even become invalid (e.g. `cos(a)` is valid but `cos(b)` is not).
</blockquote></details>

We can't add a check here since this is a Replace function. We could add the check in `build_logical_properties.go` when we determine the equivalent columns, but I thought invalid comparisons would already be caught by the type checking code in the `optbuilder` and thrown as an error? 

Based on your earlier suggestion, I also updated `build_logical_properties.go` to only save equivalences for columns when neither one has a composite key encoding -- which excludes collated strings, floats and decimals. Do you think that covers the edge cases you're thinking of?

---

*[pkg/sql/opt/norm/rules/join.opt, line 57 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis3-r7-57:-LG1WZQp-R3Nh-ARfvxj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/rules/join.opt#L57)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Add line breaks to make this more readable. I'd typically do this:
```
(InnerJoin | InnerJoinApply | SemiJoin | SemiJoinApply
    $left:*
    $right:*
    $filters:(Filters
        $list:[
            ...
            $condition:* &
                ^(Eq (Variable) (Variable)) &
                (CanMap $filters $condition $left) &
                (CanMap $filters $condition $right)
            ...
        ])
)
```
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/rules/join.opt, line 63 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis3-r7-63:-LG1WiLs7sz29T4fpf3a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/rules/join.opt#L63)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd use the list constructor syntax here:
```
(Filter [(Map $filters $condition $left)])
```
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/rules/join.opt, line 91 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis3-r7-91:-LG1XRLY6L3yTgZyVUCp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/rules/join.opt#L91)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Same line break comment as above.
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/rules/join.opt, line 108 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis3-r7-108:-LG1XRnXDj3Fe-UOTzd9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/rules/join.opt#L108)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Same line break comment as above.
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/rules/select.opt, line 110 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis4-r7-110:-LG1XVvrDc0B61dCjt4O:b-iyfftg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/rules/select.opt#L110)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Same comments as above (line break + use list constructor syntax).
</blockquote></details>

This rule was removed.

---

*[pkg/sql/opt/norm/rules/select.opt, line 194 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIis4-r7-194:-LG1XbL3BraoAjFFz-Bq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/rules/select.opt#L194)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we should just move `MergeSelectInnerJoin` above the `PushSelectIntoJoin` rules to replace this rule. Continuing to keep it below those rules comes at the cost of extra complexity that I don't think is worth it. Most joins will have an equality, so I don't think it saves us much anyway. We can remove `HasEquivalence` as well.
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/testdata/rules/join, line 707 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIirm-r7-707:-LG1_a5D-CntBvU8nK3B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/testdata/rules/join#L707)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How about a test case that involves equality conditions that only involve columns from one side or the other? The idea is to try and confuse the rules with a non-join equality.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/testdata/external/tpch, line 353 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFhjRwXESo_Ok3iIirx-r6-353:-LG1_bhjBsfyLnynoFXE:btojbfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3c0e05c7d2108e3cf633e6e940a1b6213edee36/pkg/sql/opt/xform/testdata/external/tpch#L353)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm working on an FDSet change to get rid of "removed" by normalizing the FDs. That will help us simplify in cases like this.
</blockquote></details>

Great!

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/norm/custom_funcs.go, line 435 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFxEpio3lXOMMo9UCdd:-LG1fhHw7CqZy-x-UUtm:bnybrvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L435)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

We can't add a check here since this is a Replace function. We could add the check in `build_logical_properties.go` when we determine the equivalent columns, but I thought invalid comparisons would already be caught by the type checking code in the `optbuilder` and thrown as an error? 

Based on your earlier suggestion, I also updated `build_logical_properties.go` to only save equivalences for columns when neither one has a composite key encoding -- which excludes collated strings, floats and decimals. Do you think that covers the edge cases you're thinking of?
</blockquote></details>

It's valid to compare an int to float..

Yeah, the change you made should cover the cases with composite things. We may want to revisit if we do a poor job at optimizing such queries (btw I recently noticed that even floats can be composite because of `-0`)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/opt/norm/custom_funcs.go, line 435 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFxEpio3lXOMMo9UCdd:-LG1jHXB1nWJrp2VF9fT:bkezphp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L435)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Sounds good.
</blockquote></details>

To be clear, I still think we need to prevent var substitution if the types don't match. Here is a concrete example (since the `cos` example is fixed by the otherwise unrelated composite check):
```
CREATE TABLE t1 (a DATE);
CREATE TABLE t2 (b TIMESTAMPTZ);

SELECT * FROM t1, t2 WHERE a = b AND AGE(b, TIMESTAMPTZ '2017-01-01') > INTERVAL '1 day';
+---------------------------+---------------------------+
|             a             |             b             |
+---------------------------+---------------------------+
| 2018-01-01 00:00:00+00:00 | 2018-01-01 00:00:00+00:00 |
+---------------------------+---------------------------+

SELECT * FROM t1, t2 WHERE a = b AND AGE(a, TIMESTAMPTZ '2017-01-01') > INTERVAL '1 day';
pq: unknown signature: age(date, timestamptz) (desired <interval>)
```
It is not valid to substitute `b` with `a` in that expression, even though they are equal. I would add this as a test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/norm/custom_funcs.go, line 435 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFxEpio3lXOMMo9UCdd:-LG1l-Pw8nYXWcDsJRS9:b-7ampks) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L435)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

To be clear, I still think we need to prevent var substitution if the types don't match. Here is a concrete example (since the `cos` example is fixed by the otherwise unrelated composite check):
```
CREATE TABLE t1 (a DATE);
CREATE TABLE t2 (b TIMESTAMPTZ);

SELECT * FROM t1, t2 WHERE a = b AND AGE(b, TIMESTAMPTZ '2017-01-01') > INTERVAL '1 day';
+---------------------------+---------------------------+
|             a             |             b             |
+---------------------------+---------------------------+
| 2018-01-01 00:00:00+00:00 | 2018-01-01 00:00:00+00:00 |
+---------------------------+---------------------------+

SELECT * FROM t1, t2 WHERE a = b AND AGE(a, TIMESTAMPTZ '2017-01-01') > INTERVAL '1 day';
pq: unknown signature: age(date, timestamptz) (desired <interval>)
```
It is not valid to substitute `b` with `a` in that expression, even though they are equal. I would add this as a test.

</blockquote></details>

Ah ok thanks for the example! I'll add a check

---

*[pkg/sql/opt/norm/rules/join.opt, line 57 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFscxbk4VMJZMpXL1vz:-LG1hjsS9Dg0tDKM-cEn:b-fzks9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/978fe10f1f8b875cfbf135f21eaf18cffc56e87f/pkg/sql/opt/norm/rules/join.opt#L57)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

On the `(Eq Var Var)` exception: we could have something like `left.a=right.a AND left.a=right.b` and it would be nice to push that down as `right.a=right.b`
</blockquote></details>

I have an idea about how to add these additional equality filters, but it's a bit complicated so I'd rather wait until the next PR. There are few complications:
1. We want to make sure the same rule doesn't get applied over and over
2. We don't want to lose the information about equal columns until all the other filters have been pushed down
3. We should push all the equality filters down at once in case there is a larger equality group (e.g., `left.a=right.a AND left.a=right.b AND right.b=left.c`)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Ok, I think I've made all the changes. I moved the logic to check composite types as well as equivalent types into `CanMap`, since I think it's good to keep the FD calculation simple, and it's relatively cheap to check the type stuff. PTAL.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/norm/custom_funcs.go, line 435 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFxEpio3lXOMMo9UCdd:-LG24OpT7Fm2Woz79pP8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d99ce7a5d0205129f3f0551c6a335358cf276346/pkg/sql/opt/norm/custom_funcs.go#L435)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Ah ok thanks for the example! I'll add a check
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LG23lYwAn5_zWoLhKFT:b-nssqzm)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks for all this work! The rules look good to me. I'd wait for Andy to take a look at the final version too though.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/norm/custom_funcs.go, line 433 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5PtFf4iUnyHa0QVbI:-LG5PtFf4iUnyHa0QVbJ:b-t7g1pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/custom_funcs.go#L433)):*
> ```Go
> 		found := false
> 		for j, ok := eqCols.Next(0); ok; j, ok = eqCols.Next(j + 1) {
> 			dstColType := c.f.Metadata().ColumnType(opt.ColumnID(j))
> ```

[nit] I would extract this check in a separate function because it's imperative that both CanMap and Map use the same check. I would also move the `DatumTypeHasCompositeKeyEncoding` check in there for clarity, even if it's slightly less efficient.

---

*[pkg/sql/opt/norm/testdata/rules/join, line 927 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5QFzl2Z9HXc0gLovb:-LG5QFzl2Z9HXc0gLovc:b8na154) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/testdata/rules/join#L927)):*
> ```
> # with different types.
> opt
> SELECT * FROM t1, t2 WHERE a = b AND AGE(b, TIMESTAMPTZ '2017-01-01') > INTERVAL '1 day'
> ```

Could you also add this as a logic test? I think it's valuable to have since it's less likely to accept a diff there; I'd insert a date and a timestamp that are equal to verify that we get the result.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LG5WCW771iHTpxPi4LG:b-by22d6)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/memo/expr_view.go, line 442 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFsTVdk8MwskZjvLvE6:-LFsTVdk8MwskZjvLvE7:b-xjl5ex) ([raw file](https://github.com/cockroachdb/cockroach/blob/cee2bb766d3fb7d03e0a4d4f0182d41fe3022504/pkg/sql/opt/memo/expr_view.go#L442)):*
> ```Go
> 		hasFuncDeps := !f.HasFlags(opt.ExprFmtHideFuncDeps) && !scalar.FuncDeps.Empty()
> 
> 		if showType || hasOuterCols || hasConstraints {
> ```

I think you need `|| hasFuncDeps` as well here.

---

*[pkg/sql/opt/norm/custom_funcs.go, line 420 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5fTWY21iZTsfioNIX:-LG5fTWY21iZTsfioNIY:buxid8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/custom_funcs.go#L420)):*
> ```Go
> 		// TODO(rytaft): In the future, we may want to allow the mapping if the
> 		// filter involves a comparison operator, such as x < 5.
> 		srcColType := c.f.Metadata().ColumnType(opt.ColumnID(i))
> ```

I think it would be cleaner to move this type equivalence code into `logicalPropsBuilder`. I don't think we should even add equivalence dependencies unless the columns are equivalent according to these definitions. Once we do that, the `CanMap` and `Map` functions should get much simpler. This would also address Radu's comment re: consolidating the type mapping code.

---

*[pkg/sql/opt/norm/testdata/rules/select, line 413 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5ll3q05bLfPJk1cDK:-LG5ll3q05bLfPJk1cDL:bmd7hsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/testdata/rules/select#L413)):*
> ```
> 
> opt
> SELECT * FROM a LEFT JOIN xy ON a.k=xy.x WHERE a.f=1.1 AND a.i<xy.y AND (a.s='foo' OR a.s='bar')
> ```

You should change this test case and the others below so that they're testing what we intended to test. We don't want these outer joins transformed into inner joins, or the original rules won't be triggered.  Maybe you can add `OR <var> IS NULL` clauses to some of the predicates to prevent the new rules from firing.

---

*[pkg/sql/opt/norm/testdata/rules/select, line 794 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5m8tS7JAVuqeD9TZT:-LG5m8tS7JAVuqeD9TZU:b-t6y5vi) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/testdata/rules/select#L794)):*
> ```
>  └── false [type=bool]
> 
> # Don't merge with LEFT JOIN.
> ```

More tests that needs to be updated to avoid the new rules...

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the reviews! Updated.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/memo/expr_view.go, line 442 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LFsTVdk8MwskZjvLvE6:-LG6U8YHCNMGD5xn7Zxs:bt3f2em) ([raw file](https://github.com/cockroachdb/cockroach/blob/cee2bb766d3fb7d03e0a4d4f0182d41fe3022504/pkg/sql/opt/memo/expr_view.go#L442)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think you need `|| hasFuncDeps` as well here.
</blockquote></details>

It's there -- are you looking at the latest version?

---

*[pkg/sql/opt/norm/custom_funcs.go, line 420 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5fTWY21iZTsfioNIX:-LG76KM3BjUFMXHY2mnt:bmyvph4) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/custom_funcs.go#L420)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think it would be cleaner to move this type equivalence code into `logicalPropsBuilder`. I don't think we should even add equivalence dependencies unless the columns are equivalent according to these definitions. Once we do that, the `CanMap` and `Map` functions should get much simpler. This would also address Radu's comment re: consolidating the type mapping code.
</blockquote></details>

Based on our offline discussion I've created a helper function

---

*[pkg/sql/opt/norm/custom_funcs.go, line 433 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5PtFf4iUnyHa0QVbI:-LG76N1N4I9megvCVIsQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/custom_funcs.go#L433)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I would extract this check in a separate function because it's imperative that both CanMap and Map use the same check. I would also move the `DatumTypeHasCompositeKeyEncoding` check in there for clarity, even if it's slightly less efficient.
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/testdata/rules/join, line 927 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5QFzl2Z9HXc0gLovb:-LG7D2e37UI-pdGMnh2r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/testdata/rules/join#L927)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Could you also add this as a logic test? I think it's valuable to have since it's less likely to accept a diff there; I'd insert a date and a timestamp that are equal to verify that we get the result.
</blockquote></details>

Done.

---

*[pkg/sql/opt/norm/testdata/rules/select, line 413 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5ll3q05bLfPJk1cDK:-LG7FxmZDD707zGn4jgi:b-376vyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/testdata/rules/select#L413)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

You should change this test case and the others below so that they're testing what we intended to test. We don't want these outer joins transformed into inner joins, or the original rules won't be triggered.  Maybe you can add `OR <var> IS NULL` clauses to some of the predicates to prevent the new rules from firing.
</blockquote></details>

Done. Good catch.

---

*[pkg/sql/opt/norm/testdata/rules/select, line 794 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-LG5m8tS7JAVuqeD9TZT:-LG7Fw1J1IBWVJ0WRBeE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0af2572cf7fedec3ad922f293492efed0b39593f/pkg/sql/opt/norm/testdata/rules/select#L794)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

More tests that needs to be updated to avoid the new rules...
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LG7I7ft2UxsXq9F83Qc:bc7d8fp)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LG7gYLCDYLDI4PVmu8v:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
bors r+

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LG8nDwGAe991xxSXCPF:b-ydi9ay)*
<!-- Sent from Reviewable.io -->

rytaft commented:
bors r+

---







*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LGAP-q4Emi9fOtxVk6w:b-ydi9ay)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

rytaft commented:
bors r+

---







*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26937#-:-LGBA8yc-fha8gBi-ZUG:b-ydi9ay)*
<!-- Sent from Reviewable.io -->
